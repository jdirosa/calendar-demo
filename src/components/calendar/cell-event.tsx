import React from 'react';
import { CalendarEvent, EventType } from '../../types';
import { makeStyles } from '@material-ui/styles';
import { EVENT_COLORS } from '../../constants';
import { trimText } from '../../utils/text-helpers';
import { Typography } from '@material-ui/core';

const styles = (event: CalendarEvent) =>
  makeStyles(theme => ({
    event: {
      margin: 10,
      padding: 6,
      backgroundColor:
        event.type === EventType.Birthday
          ? EVENT_COLORS.birthday
          : event.type === EventType.Holiday
          ? EVENT_COLORS.holiday
          : event.type === EventType.NewHire
          ? EVENT_COLORS.newHire
          : EVENT_COLORS.personalDay,
      color: 'white',
      borderRadius: 5,
    },
  }));

interface Props {
  event: CalendarEvent;
}
export const EventMarker: React.FunctionComponent<Props> = ({ event }) => {
  const classes = styles(event)();
  const text = trimText(16, event.title);
  return (
    <div className={classes.event}>
      <Typography>{text}</Typography>
    </div>
  );
};
