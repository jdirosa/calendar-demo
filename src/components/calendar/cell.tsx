import React from 'react';
import { Grid, makeStyles, Typography, Avatar } from '@material-ui/core';
import { weekdays, months, areDatesOnSameDay } from '../../utils/date-helpers';
import { useEventContext } from '../../context-providers/event-context';
import { EventMarker } from './cell-event';

interface Props {
  header: boolean;
  date: Date;
  onClick: () => void;
}

const cellStyles = makeStyles(theme => ({
  grid: {
    width: 'calc(100% / 7)',
  },
  paper: {
    borderRight: '1px solid #ccc',
    borderBottom: '1px solid #ccc',
    padding: 6,
    borderRadius: 0,
    minHeight: 140,
    textAlign: 'center',
    color: theme.palette.text.secondary,
  },
  today: {
    backgroundColor: 'rgba(222, 53, 158,1)',
    color: 'white',
    fontSize: 14,
    width: 30,
    height: 30,
  },
  avatar: {
    backgroundColor: 'white',
    color: '#333',
    fontSize: 14,
  },
}));
export const Cell: React.FunctionComponent<Props> = ({
  date,
  header,
  onClick,
}) => {
  const classes = cellStyles();
  const day = date.getDate();
  const dateText = day === 1 ? `${months[date.getMonth()]} ${day}` : day;
  const ctx = useEventContext();
  const event = ctx.getEvent(date);
  const isToday = areDatesOnSameDay(new Date(), date);
  return (
    <Grid item className={classes.grid}>
      <div className={classes.paper} onClick={onClick}>
        {header && (
          <Typography style={{ marginBottom: 6 }} variant="h6" component="h4">
            {weekdays[date.getDay()]}
          </Typography>
        )}
        {day === 1 ? (
          <Typography component="div">{dateText}</Typography>
        ) : (
          <Avatar className={isToday ? classes.today : classes.avatar}>
            {dateText}
          </Avatar>
        )}

        {event && <EventMarker event={event} />}
      </div>
    </Grid>
  );
};
