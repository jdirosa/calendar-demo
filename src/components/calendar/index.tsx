import React from 'react';
import { makeStyles, Grid, Paper, Typography } from '@material-ui/core';
import { Cell } from './cell';
import {
  getDaysInMonth,
  getNextDate,
  getCalendarOffset,
  months,
} from '../../utils/date-helpers';
import { NewEvent } from './modals/new-event';
import { useEventContext } from '../../context-providers/event-context';
import { EventType } from '../../types';

const useStyles = makeStyles(theme => ({
  root: {
    flexGrow: 1,
    marginTop: 20,
  },
}));
export const Calendar: React.FunctionComponent = () => {
  const classes = useStyles();
  const [selectedDate, setSelectedDate] = React.useState<Date | undefined>();
  const ctx = useEventContext();
  const handleSaveEvent = (
    title: string,
    description: string,
    type: EventType
  ) => {
    if (!selectedDate) {
      return;
    }
    ctx.addEvent(title, description, type, selectedDate);
    setSelectedDate(undefined);
  };
  const handleCellClick = (date: Date) => {
    const event = ctx.getEvent(date);
    if (event) {
      ctx.setActiveEvent(event);
    } else {
      setSelectedDate(date);
    }
  };
  const currentMonth = months[new Date().getMonth()];
  return (
    <React.Fragment>
      <Typography variant="h4">{currentMonth}</Typography>
      <Paper className={classes.root}>
        <Grid container>
          {getCalendarDays().map((d, i) => (
            <Cell
              key={i}
              date={getNextDate(d)}
              header={i < 7}
              onClick={() => handleCellClick(getNextDate(d))}
            />
          ))}
        </Grid>
        <NewEvent
          date={selectedDate}
          onSave={handleSaveEvent}
          onCancel={() => setSelectedDate(undefined)}
        />
      </Paper>
    </React.Fragment>
  );
};

/** Returns the number of days required to perfectly balance the calendar for the month. Including previous month and next month days */
const getCalendarDays = () => {
  const d = [];
  const offset = getCalendarOffset();

  const totalDays = getDaysInMonth() + offset;
  for (let i = -1 * offset; i < totalDays; i++) {
    d.push(i);
  }
  return d.sort((a, b) => a - b);
};
