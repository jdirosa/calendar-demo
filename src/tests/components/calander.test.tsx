import renderer from 'react-test-renderer';
import React from 'react';
import { Calendar } from '../../components/calendar';

it('renders correctly', () => {
  const cal = renderer.create(<Calendar />).toJSON();
  expect(cal).toMatchSnapshot();
});
