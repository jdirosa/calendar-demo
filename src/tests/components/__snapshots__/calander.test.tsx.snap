// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <h4
    className="MuiTypography-root MuiTypography-h4"
  >
    September
  </h4>,
  <div
    className="MuiPaper-root MuiPaper-elevation1 makeStyles-root-1 MuiPaper-rounded"
  >
    <div
      className="MuiGrid-root MuiGrid-container"
    >
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <h4
            className="MuiTypography-root MuiTypography-h6"
            style={
              Object {
                "marginBottom": 6,
              }
            }
          >
            Monday
          </h4>
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            26
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <h4
            className="MuiTypography-root MuiTypography-h6"
            style={
              Object {
                "marginBottom": 6,
              }
            }
          >
            Tuesday
          </h4>
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            27
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <h4
            className="MuiTypography-root MuiTypography-h6"
            style={
              Object {
                "marginBottom": 6,
              }
            }
          >
            Wednesday
          </h4>
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            28
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <h4
            className="MuiTypography-root MuiTypography-h6"
            style={
              Object {
                "marginBottom": 6,
              }
            }
          >
            Thursday
          </h4>
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            29
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <h4
            className="MuiTypography-root MuiTypography-h6"
            style={
              Object {
                "marginBottom": 6,
              }
            }
          >
            Friday
          </h4>
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            30
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <h4
            className="MuiTypography-root MuiTypography-h6"
            style={
              Object {
                "marginBottom": 6,
              }
            }
          >
            Saturday
          </h4>
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            31
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <h4
            className="MuiTypography-root MuiTypography-h6"
            style={
              Object {
                "marginBottom": 6,
              }
            }
          >
            Sunday
          </h4>
          <div
            className="MuiTypography-root MuiTypography-body1"
          >
            September 1
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            2
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            3
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            4
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-today-164 MuiAvatar-colorDefault"
          >
            5
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            6
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            7
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            8
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            9
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            10
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            11
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            12
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            13
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            14
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            15
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            16
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            17
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            18
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            19
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            20
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            21
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            22
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            23
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            24
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            25
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            26
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            27
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            28
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            29
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            30
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiTypography-root MuiTypography-body1"
          >
            October 1
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            2
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            3
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            4
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            5
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-grid-162 MuiGrid-item"
      >
        <div
          className="makeStyles-paper-163"
          onClick={[Function]}
        >
          <div
            className="MuiAvatar-root makeStyles-avatar-165 MuiAvatar-colorDefault"
          >
            6
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
